### ISSUE TITLE
Animate ConstraintSet changes doesn't take into account the animationSpec

### ISSUE THREAD

**Issue by riggaroo** [androidx/constraintlayout#655]
When animating between two constraintSets, it doesn't seem to take into account the specified animationSpec - if I increase the duration or change the damping ratio, the animation doesn't change. 

```
@Composable
fun ConstraintLayoutAnim() {
    var vertical by remember {
        mutableStateOf(true)
    }

    val constraintSet = remember(vertical) {
        ConstraintSet {
            val button = createRefFor("button")          // Create ID "button"
            val title = createRefFor("title")            // Create ID "title"
            val g1 = createGuidelineFromStart(80.dp)     // Create Guideline "g1"
            constrain(button) {
                if (vertical) {
                    top.linkTo(title.bottom, 16.dp)          // Constrain top to bottom of title with 16dp margin
                    start.linkTo(g1)                         // Constrain start to g1 (0dp margin implied)
                    // Create Constraint for id "button"
                } else {
                    start.linkTo(title.end, 16.dp)
                    top.linkTo(title.top)
                    bottom.linkTo(title.bottom)
                }
            }
            constrain(title) {                           // Create Constraint for id "title"
                centerVerticallyTo(parent)               // center vertically in parent (equivalent to top_toTop="parent" bottom_toBottom="parent"
                start.linkTo(g1)                         // Constrain start to guideline
            }
        }
    }

    // doesn't work as expected
    ConstraintLayout(
        constraintSet,
        modifier = Modifier.fillMaxSize(),
        animateChanges = true,
        animationSpec = spring(dampingRatio = Spring.DampingRatioHighBouncy, stiffness = Spring
            .StiffnessVeryLow)
    ) {
        Button(
            modifier = Modifier.layoutId("button"),      // Button has an id "button"
            onClick = { vertical = !vertical},
        ) {
            Text(text = "Hi")
        }
        Text(modifier = Modifier.layoutId("title"),       // Text has an id "title"
            text = "Welcome"
        )
    }
}


```
It also seems to get slower on each press of the button. 

[anim_issue.webm](https://user-images.githubusercontent.com/9973046/181211617-a9fd98af-157c-4b96-8741-f47a4762db21.webm)

I was expecting to see my buttons bounce around the screen when specifying `animateChanges = true` with a `spring` animationSpec.

**Comment by jafu888** [comment]
It only works if you switch constraintSets not if you modify one constraintSet
See example [AnimatedConstraintssExample ](https://github.com/androidx/constraintlayout/blob/17896f915b3058370c681834f4f246381136f050/projects/ComposeConstraintLayout/app/src/main/java/com/example/constraintlayout/MotionComposeExamples.kt#L55)

 So in your case something like :
 
 ```kotlin
 
fun layout(vertical : Boolean) : ConstraintSet {
    return   ConstraintSet {
        val button = createRefFor("button")          // Create ID "button"
        val title = createRefFor("title")            // Create ID "title"
        val g1 = createGuidelineFromStart(80.dp)     // Create Guideline "g1"
        constrain(button) {
            if (vertical) {
                top.linkTo(title.bottom, 16.dp)          // Constrain top to bottom of title with 16dp margin
                start.linkTo(g1)                         // Constrain start to g1 (0dp margin implied)
                // Create Constraint for id "button"
            } else {
                start.linkTo(title.end, 16.dp)
                top.linkTo(title.top)
                bottom.linkTo(title.bottom)
            }
        }
        constrain(title) {                           // Create Constraint for id "title"
            centerVerticallyTo(parent)               // center vertically in parent (equivalent to top_toTop="parent" bottom_toBottom="parent"
            start.linkTo(g1)                         // Constrain start to guideline
        }
    }
}

@Preview
@Composable
fun ConstraintLayoutAnim() {
    var vertical by remember {
        mutableStateOf(true)
    }

    val constraintSet1 = remember() { layout(true); }
    val constraintSet2 = remember() { layout(false); }
    
    ConstraintLayout(
        if (vertical) constraintSet1 else constraintSet2,
        modifier = Modifier.fillMaxSize(),
        animateChanges = true,
        animationSpec = spring(dampingRatio = Spring.DampingRatioHighBouncy, stiffness = Spring
            .StiffnessVeryLow)
    ) {
        Button(
            modifier = Modifier.layoutId("button"),      // Button has an id "button"
            onClick = { vertical = !vertical},
        ) {
            Text(text = "Hi")
        }
        Text(modifier = Modifier.layoutId("title"),       // Text has an id "title"
            text = "Welcome"
        )
    }
}

 ```

**Comment by jafu888** [comment]
![cl_set_change](https://user-images.githubusercontent.com/15019413/181375907-9a288ba7-a424-4c6a-b6bf-152b808254fb.gif)

**Comment by oscar-ad** [comment]
Just to clarify what is happening. `ConstrainSet` lambdas are evaluated at measure time, not at composition (this is a mistake, we eventually want to make it friendlier with the Compose cycle).

So a `ConstraintSet` that depends on a mutable value will reflect the updated value during animation. To support `animateChanges` we basically save the old ConstraintSet as the `start` ConstraintSet; and the new one is considered the `end` ConstraintSet. 
In this case, because of how they are evaluated, both ConstraintSets result in the same layout (with the updated state).

In other words, when animating, `ConstraintSet`s should only take immutable (and state-less) values.

**Comment by riggaroo** [comment]
Cool - thanks for the clarification. Will close this as working as expected then.
